### PR TITLE
add a Handle::current() for getting access to TLS Runtime

### PR DESCRIPTION
**Pull request by bluejekyll**
## Motivation

Get a Runtime::Handle to the current thread local storage Runtime components.

## Solution

This adds a blocking::Spawner reference to ThreadContext. With that we then have access to all the components needed for a Handle::current. This will panic if components haven't been registered that are required.

### CODE REVIEW

File: tokio/src/runtime/blocking/mod.rs [hunk 1]
```diff
@@ -60,6 +60,8 @@ cfg_not_blocking_impl! {
         where
             F: FnOnce() -> R,
         {
+            let ctx = crate::runtime::context::ThreadContext::clone_current();
```
[the remark this answers is not in the document]
Comment (carllerche) [reply tokio/src/runtime/blocking/mod.rs]: Before, `ThreadContext` is entered [here](https://github.com/tokio-rs/tokio/blob/e43f28f6a8290d8b452c707de06e2d95236f59e3/tokio/src/runtime/blocking/pool.rs#L255) and the blocking pool context is set [here](https://github.com/tokio-rs/tokio/blob/e43f28f6a8290d8b452c707de06e2d95236f59e3/tokio/src/runtime/blocking/pool.rs#L264) (not an ideal setup, but it worked so far).

File: tokio/src/runtime/context.rs [hunk 2]
```diff
@@ -93,31 +99,54 @@ impl ThreadContext {
...
 
-    #[cfg(all(feature = "io-driver", not(loom)))]
+    #[cfg(any(
+        feature = "blocking",
+        feature = "dns",
+        feature = "fs",
+        feature = "io-std",
+        feature = "rt-threaded",
+    ))]
+    pub(crate) fn with_blocking_spawner(
```
Comment (carllerche) [inline tokio/src/runtime/context.rs]: I think this can be replaced w/ the `cfg_blocking_impl` macro ([here](https://github.com/tokio-rs/tokio/blob/master/tokio/src/macros/cfg.rs#L24)). It cannot go around fn defs, so you would split it out to another impl block:

```rust
cfg_blocking_impl! {
    impl ThreadContext {
        pub(crate) fn with_blocking_spawner(....) { ... }
    }
}
```

File: tokio/src/runtime/context.rs [hunk 3]
```diff
@@ -93,31 +99,54 @@ impl ThreadContext {
...
+        mut self,
+        blocking_spawner: crate::runtime::blocking::Spawner,
+    ) -> Self {
+        self.blocking_spawner.replace(blocking_spawner);
+        self
+    }
+
     pub(crate) fn io_handle() -> crate::runtime::io::Handle {
+        #[cfg(any(not(feature = "io-driver"), loom))]
```
Comment (carllerche) [inline tokio/src/runtime/context.rs]: Could you elaborate on why this change was made?

File: tokio/src/runtime/handle.rs [hunk 4]
```diff
@@ -35,9 +35,30 @@ impl Handle {
...
+            Some(self.blocking_spawner.clone()),
         )
         .enter();
-        self.blocking_spawner.enter(|| f())
+
+        f()
+    }
+
+    /// Returns a Handle view over the currently running Runtime
+    ///
```
Comment (carllerche) [inline tokio/src/runtime/handle.rs]: Could I trouble you for a blurb on why/when you would need to use this function? Also, maybe a quick doc example? :heart:

**Comment by carllerche** [comment]
This looks good to me. I added some minor inline comments.

File: tokio/src/runtime/blocking/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bluejekyll) [reply tokio/src/runtime/blocking/mod.rs]: Ok. Would you like me to remove this?

File: tokio/src/runtime/context.rs (hunk 3 above)
Comment (bluejekyll) [reply tokio/src/runtime/context.rs]: Oh. I think it was failing CI with a unused warning? Though I can try reverting it.

File: tokio/src/runtime/blocking/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (carllerche) [reply tokio/src/runtime/blocking/mod.rs]: I think removing it would be best.

File: tokio/src/runtime/context.rs (hunk 3 above)
Comment (carllerche) [reply tokio/src/runtime/context.rs]: Should the `fn io_handle` be scoped w/ a `cfg` then? The fns below have had their `cfg` bits removed. I'd be interested in seeing the error.

Comment (bluejekyll) [reply tokio/src/runtime/context.rs]: So the way that it's written right now, the field is not optional, it gets swapped out for `()` when these features aren't enabled (apparently). Here's the error with a default cargo check with those changed reverted::

```shell
error[E0308]: match arms have incompatible types
   --> tokio/src/runtime/context.rs:105:21
    |
103 |           CONTEXT.with(|ctx| match *ctx.borrow() {
    |  ____________________________-
104 | |             Some(ref ctx) => ctx.io_handle.clone(),
    | |                              --------------------- this is found to be of type `()`
105 | |             None => None,
    | |                     ^^^^ expected (), found enum `std::option::Option`
106 | |         })
    | |_________- `match` arms have incompatible types
    |
    = note: expected type `()`
               found type `std::option::Option<_>`
```

Comment (carllerche) [reply tokio/src/runtime/context.rs]: Ah! What if you sub `None` for `Default::default()`?

That said, it is curious that `io_handle()` is required even when the `io-driver` feature isn't enabled.

Comment (bluejekyll) [reply tokio/src/runtime/context.rs]: This stems from the way the Handle::Current works, which asks for all the components of ThreadContext, where Handle has all of these, so cfging off the io_handle function will spider out for a lot of other changes. Maybe Default::default() will work...

Comment (bluejekyll) [reply tokio/src/runtime/context.rs]: Ok, that works.

**Comment by bluejekyll** [comment]
ok, @carllerche , I think I cleaned up what could be. Let me know if you want me to take care of anything else.

File: tokio/src/runtime/handle.rs (hunk 4 above)
Comment (bluejekyll) [reply tokio/src/runtime/handle.rs]: Is the new doc enough?

File: tokio/src/runtime/context.rs (hunk 2 above)
Comment (bluejekyll) [reply tokio/src/runtime/context.rs]: Ok, done.

File: tokio/src/runtime/blocking/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bluejekyll) [reply tokio/src/runtime/blocking/mod.rs]: Removed.

**Comment by carllerche** [comment]
I merged master & resolved the conflict.

**Comment by bluejekyll** [comment]
Thanks! I didn't notice we ended up with a conflict.

**Comment by carllerche** [comment]
I've been merging a lot of PRs

File: tokio/src/runtime/handle.rs (hunk 4 above)
Comment (carllerche) [reply tokio/src/runtime/handle.rs]: Looks great :+1:

**Comment by bluejekyll** [comment]
No problem. It's really benefiting from all the work that was done on ThreadContext.

Thanks for managing all of this!
